src/database_abstraction_layer.py

```diff
@@ -0,0 +1,153 @@
+
+log = get_logger()
+s3 = boto3.client("s3")
+
+
+class DataAbstractionLayer:
+    # Database paths and URLs can be set here
+    S3_BUCKET = "s3://amos--data--events/leads/enriched.csv"
+    LOCAL_OUTPUT = "./data/leads_enriched.csv"
+
+    def __init__(self, db_type=None):
+        """
+        Initialise DAL
+        :param db_type: Database to be used for output. As of 04/12/23, it can be: S3, local
+        """
+        # Attributes for all data types
+        self.db_type = db_type
+        self.df = None
+
+        # S3
+        self.bucket = None
+        self.obj_key = None
+
+        if self.db_type == "S3" or self.db_type == "local":
+            # Use S3 as the input location for both S3 and local
+            self._connect_s3()
+        elif self.db_type == None:
+            raise log.error("No output location selected.")
+        elif self.db_type is not None:
+            raise log.error("Unsupported database type")
+
+    def _connect_s3(self):
+        """
+        Download database from specified S3 file
+        """
+        if not self.S3_BUCKET.startswith("s3://"):
+            log.error(
+                "S3 location has to be defined like this: s3://<BUCKET>/<OBJECT_KEY>"
+            )
+            return
+
+        source = self.S3_BUCKET
+        remote_dataset = None
+
+        try:
+            self.bucket, self.obj_key = self._decode_s3_url()
+            remote_dataset = self._fetch_object_s3()
+        except IndexError:
+            log.error(
+                "S3 location has to be defined like this: s3://<BUCKET>/<OBJECT_KEY>"
+            )
+
+        if remote_dataset is None or "Body" not in remote_dataset:
+            log.error(
+                f"Couldn't find dataset in S3 bucket {self.bucket} and key {self.obj_key}"
+            )
+            return
+        else:
+            source = remote_dataset["Body"]
+
+        try:
+            self.df = pd.read_csv(source)
+        except FileNotFoundError:
+            log.error("Error: Could not find input file for Pipeline.")
+
+    def _fetch_object_s3(self):
+        """
+        Tries to read an object from S3.
+        :return: s3 object
+        """
+        obj = None
+        try:
+            obj = s3.get_object(Bucket=self.bucket, Key=self.obj_key)
+        except botocore.exceptions.ClientError as e:
+            log.warning(
+                f"{e.response['Error']['Code']}: {e.response['Error']['Message']}"
+                if "Error" in e.response
+                else f"Error while getting object s3://{self.bucket}/{self.obj_key}"
+            )
+
+        return obj
+
+    def _decode_s3_url(self):
+        """
+        Retrieve the bucket and object key from object url
+        :return: bucket string, object key string
+        """
+        obj_identifier = self.S3_BUCKET.split("//")[1].split("/")
+        bucket = obj_identifier[0]
+        obj_key = "/".join(obj_identifier[1:])
+        return bucket, obj_key

```

Comment:
Does this function work with more nested folders?

Reply:
I think it should, because you can't have buckets within buckets, but you can have folders within folders. The entire path after the bucket is the object key.

Reply:
Yeah, the way André explained it is that S3 doesn't really know folders. It will just interpret the object key as a folder structure if you add "/" characters and display it in the UI.